node/src/store/header_ranges.rs

```diff
@@ -0,0 +1,566 @@
+use std::fmt::Display;
+use std::ops::RangeInclusive;
+use std::vec;
+
+#[cfg(any(test, feature = "test-utils"))]
+use celestia_types::test_utils::ExtendedHeaderGenerator;
+use celestia_types::ExtendedHeader;
+use serde::Serialize;
+use smallvec::{IntoIter, SmallVec};
+
+use crate::store::utils::{ranges_intersection, try_consolidate_ranges, RangeScanResult};
+use crate::store::StoreError;
+
+pub type HeaderRange = RangeInclusive<u64>;
+
+macro_rules! header_ranges {
+    ( $( $r:expr ),* ) => {
+        {
+            $crate::store::header_ranges::HeaderRanges::from_vec(smallvec::smallvec![
+            $(
+                $r,
+            )*
+            ])
+        }
+    };
+}
+
+pub(crate) use header_ranges;
+
+/// Span of header that's been verified internally
+#[derive(Clone)]
+pub struct VerifiedExtendedHeaders(Vec<ExtendedHeader>);
+
+impl IntoIterator for VerifiedExtendedHeaders {
+    type Item = ExtendedHeader;
+    type IntoIter = vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
```

Comment:
Add also this and remove the `to_vec()` from the callers

```suggestion

impl<'a> TryFrom<&'a [ExtendedHeader]> for VerifiedExtendedHeaders {
    type Error = celestia_types::Error;

    fn try_from(value: &'a [ExtendedHeader]) -> Result<Self, Self::Error> {
        value.to_vec().try_into()
    }
}

```